### PR TITLE
feat(wmg): Add delete/reset cell types and genes

### PR DESCRIPTION
**Pull request by tihuan**
### Reviewers
**Functional:** 
@seve 
**Readability:** 
@MDunitz 
---

## Changes
1. Update `deleteSelectedGenesAndSelectedCellTypeIds()` event to no longer take payload, since data is in the state already
2. Add `tissueCellTypesFetched()` and `resetTissueCellTypes()` events to support delete and reset cell types
3. Add type `CellTypeMetadata`, which is string in `${id}~${tissue}~${name}` format. Also added corresponding serializer(`getAllSerializedCellTypeMetadata()`) and deserializer(`deserializeCellTypeMetadata ()`)
4. Update heat map to highlight selected cell types and genes, different text style for tissues, and reset style for tissues
5. Update `frontend/src/views/WheresMyGene/index.tsx` data flow. In general, reducer state now stores `selectedGenes` and `selectedCellTypeIds`, and `selectedCellTypes` is **derived state**, thus NOT needing to live in the store. Same thing for `tissuesWithDeletedCellTypes`, which is also a derived state that we cache via `useMemo()`
6. BUG FIX: The issue we had with React memory leak is due to debounced function not being cancelled, so we have that now for all debounced calls
7. Perf improvement: Split updating charts separately instead of in one single `useEffect()`, so we no longer trigger updates too frequently and unnecessarily -- we now have `useUpdateChart()`, useUpdateXAxisChart()` and `useUpdateYAxisChart()`
8. 

## QA steps (optional)
1. You can delete cell types and genes via the heat map
2. You can reset cell types by clicking its tissue
3. You should see tissue "Lung" in the row now, indicating multiple tissues support

### CODE REVIEW

File: frontend/.eslintrc.js [hunk 1]
```diff
@@ -38,6 +38,15 @@ module.exports = {
...
     "@typescript-eslint/explicit-function-return-type": "off",
+    "@typescript-eslint/no-unused-vars": [
+      "error",
+      {
+        args: "after-used",
+        argsIgnorePattern: "^_",
+        ignoreRestSiblings: false,
+        vars: "all",
+      },
+    ],
```
Comment (seve) [inline frontend/.eslintrc.js:49]: This explicit rule is just to tweak an already active rule to ignore the pattern, correct?

Might be worth adding a comment with a few examples captured by the pattern

File: frontend/src/views/WheresMyGene/common/store/actions.ts [hunk 2]
```diff
@@ -3,22 +3,15 @@
...
-}): GetActionTypeOfReducer<
+export function deleteSelectedGenesAndSelectedCellTypeIds(): GetActionTypeOfReducer<
   typeof REDUCERS["deleteSelectedGenesAndSelectedCellTypeIds"]
 > {
   return {
-    payload: {
-      cellTypeIdsToDelete,
-      genesToDelete,
-    },
+    payload: null,
```
Comment (seve) [inline frontend/src/views/WheresMyGene/common/store/actions.ts:14]: What was the payload, is now held in redux store, correct?

File: frontend/src/views/WheresMyGene/common/store/reducer.ts [hunk 3]
```diff
@@ -161,13 +175,12 @@ function toggleCellTypeIdToDelete(
...
 
-function deleteByIds<Item>(collection: Item[], collectionToDelete: Item[]) {
+function deleteByItems<Item>(collection: Item[], collectionToDelete: Item[]) {
   return collection.filter((item) => !collectionToDelete.includes(item));
 }
 
 function resetGenesToDeleteAndCellTypeIdsToDelete(
   state: State,
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   _: PayloadAction<null>
```
Comment (seve) [inline frontend/src/views/WheresMyGene/common/store/reducer.ts:184]: I see now!

Comment (seve) [reply frontend/src/views/WheresMyGene/common/store/reducer.ts:184]: Surprised this isn't already in the rule, is there a thread/issue on this?

File: frontend/src/views/WheresMyGene/common/types.ts [hunk 4]
```diff
@@ -1,7 +1,11 @@
 import { CSSProperties } from "react";
 
+export type Tissue = string;
+
 /** This object holds a cell type and its gene expressions. */
 export interface CellTypeSummary {
+  /** tissue name */
+  tissue: string;
```
Comment (seve) [inline frontend/src/views/WheresMyGene/common/types.ts:8]:
```suggestion
  tissue: Tissue;
```
Right?

**Review by seve — COMMENTED** [review]
Redux review 1st, looking at components next

File: frontend/src/views/WheresMyGene/common/store/reducer.ts (hunk 3 above)
Comment (tihuan) [reply frontend/src/views/WheresMyGene/common/store/reducer.ts:184]: Yeah I checked eslint doc and they want us to ignore `^_` via the config 🤔 

https://eslint.org/docs/rules/no-unused-vars#argsignorepattern

File: frontend/src/views/WheresMyGene/common/store/actions.ts (hunk 2 above)
Comment (tihuan) [reply frontend/src/views/WheresMyGene/common/store/actions.ts:14]: That's right!

File: frontend/.eslintrc.js (hunk 1 above)
Comment (tihuan) [reply frontend/.eslintrc.js:49]: Great idea! Done 🔥

File: frontend/src/views/WheresMyGene/common/types.ts (hunk 4 above)
Comment (tihuan) [reply frontend/src/views/WheresMyGene/common/types.ts:8]: Oh thanks for catching this. I went ahead and update all types to use `Tissue` 🙏

**Review by seve — COMMENTED** [review]
Couple more Q

File: frontend/src/views/WheresMyGene/components/HeatMap/hooks/useUpdateChart.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      UPDATE_THROTTLE_MS,
+      // (thuang): Trailing guarantees that the last call to the function will
+      // be executed
+      { trailing: true }
+    );
+  }, []);
+
+  useEffect(() => {
+    return () => throttledUpdateChart.cancel();
+  }, [throttledUpdateChart]);
```
[the remark this answers is not in the document]
Comment (tihuan) [reply frontend/src/views/WheresMyGene/components/HeatMap/hooks/useUpdateChart.ts:45]: we want to call `cancel()` in the useEffect clean up function, thus needing to return a function inside the useEffect callback: https://reactjs.org/docs/hooks-effect.html#effects-with-cleanup

The goal is to ensure that throttle with trailing doesn't trigger a setState after the component is unmounted, thus presenting a memleak and react set state error!

**Review by seve — APPROVED** [review]

**Pull request merged**